include/swift/AST/AutoDiff.h

```diff
@@ -77,79 +78,193 @@ class AutoDiffParameter {
+    : indices(indices), options(options) {}
+
+  unsigned getSourceIndex() const {
+    return indices.source;
+  }
+
+  llvm::BitVector getParameterIndices() const {
+    return indices.parameters;
+  }
+
+  bool isSeedable() const {
+    return options.contains(SILGradientFlags::Seedable);
+  }
+
+  bool isPreservingResult() const {
+    return options.contains(SILGradientFlags::PreservingResult);
+  }
+
+  bool isDelayed() const {
+    return options.contains(SILGradientFlags::Delayed);
+  }
+
+  // FIXME: The master configuration should have all three gradient options
+  // enabled, that is, the canonical gradient should return a delayed gradient
+  // function. We need to handle this here as well as within the
+  // differentiation pass.
+  static SILGradientOptions getCanonicalGradientOptions() {
+    return SILGradientFlags::Seedable | SILGradientFlags::PreservingResult;
+  }
 
   /// Returns the "master" configuration, which all variants with the same
   /// parameter indices can derive from.
   static
-  SILReverseAutoDiffConfiguration getMaster(unsigned sourceIndex,
-                                            ArrayRef<unsigned> paramIndices) {
-    return { sourceIndex, paramIndices,
-             /*seedable*/ true, /*preservingResult*/ true };
+  SILReverseAutoDiffConfiguration getMaster(SILReverseAutoDiffIndices indices) {
+    return {
+      indices,
+      getCanonicalGradientOptions()
+    };
   }
 
-  bool isEqual(const SILReverseAutoDiffConfiguration &other) const {
-    return sourceIndex == other.sourceIndex &&
-           parameterIndices.equals(other.parameterIndices) &&
-           seedable == other.seedable &&
-           preservingResult == other.preservingResult;
+  SILReverseAutoDiffConfiguration getWithCanonicalOptions() const {
+    return getMaster(indices);
   }
 
   bool isMaster() const {
-    return seedable && preservingResult;
+    return options.toRaw() == getCanonicalGradientOptions().toRaw();
+  }
+
+  bool operator==(const SILReverseAutoDiffConfiguration &other) const {
+    return indices == other.indices &&
+           options.toRaw() == other.options.toRaw();
   }
 };
 
 } // end namespace swift
 
 namespace llvm {
 
+using swift::SILReverseAutoDiffIndices;
 using swift::SILReverseAutoDiffConfiguration;
+using swift::SILGradientFlags;
+using swift::OptionSet;
 
 template<typename T> struct DenseMapInfo;
 
+template<> struct DenseMapInfo<SILReverseAutoDiffIndices> {
+  static SILReverseAutoDiffIndices getEmptyKey() {
+    return { DenseMapInfo<unsigned>::getEmptyKey(), BitVector() };
+  }
+
+  static SILReverseAutoDiffIndices getTombstoneKey() {
+    return { DenseMapInfo<unsigned>::getTombstoneKey(),
+             BitVector(sizeof(intptr_t), true) };
+  }
+
+  static unsigned getHashValue(const SILReverseAutoDiffIndices &Val) {
+    unsigned combinedHash =
+      hash_combine(~1U, DenseMapInfo<unsigned>::getHashValue(Val.source));
+    for (auto i : Val.parameters.set_bits())
```

Comment:
I think there is a way to get the words out of the BitVector - if so, it would be a LOT more efficient to hash them.

Reply:
I can't find any public API on `SmallBitVector` or `BitVector` that doesn't this, unfortunately. Given that the efficiency is on par with the old array-based index list, I think it's acceptable. Plus, in most possible cases there are less than 10 elements.